walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,92 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.*;
+import java.util.Comparator;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  public ArrayList<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+
+    ArrayList<TimeRange> availableTimeRanges = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> bookedTimeRanges = new ArrayList<TimeRange>();
+    bookedTimeRanges = getBookedTimeRanges(events, request);
+    
+    // Checks to see if there are no events and returns time range as whole day if true.
+    if (bookedTimeRanges.size() == 0){
+        TimeRange available = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true);
+        if(available.duration() >= request.getDuration()){
+            availableTimeRanges.add(available);
+        }
+        return availableTimeRanges; 
+    }
+
+    // Arrange booked time besed upon start time.
+    Collections.sort(bookedTimeRanges, TimeRange.ORDER_BY_START);
+
+    int i = 0;
+    // Checks for overlaps in booked time ranges and creates new array list with all unavailable times.
+    while (i < (bookedTimeRanges.size() - 1)){
+        TimeRange first = bookedTimeRanges.get(i);
+        TimeRange second = bookedTimeRanges.get(i+1);
+        if(first.overlaps(second)){
+            if (first.contains(second)){
+                bookedTimeRanges.remove(second);
+            } else {
+                int oldStart = first.start();
+                int newEnd = second.end();
+                
+                bookedTimeRanges.set(i, TimeRange.fromStartEnd(oldStart, newEnd, false));
+                bookedTimeRanges.remove(second);
+            }
+
+        }
+        i+=1; 
+    }
+
+    // Add time available to the available time ranges at the beginning of the day.
+    if (bookedTimeRanges.size() > 0){
+        if(!(bookedTimeRanges.get(0).contains(TimeRange.START_OF_DAY))){
+            TimeRange start = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, bookedTimeRanges.get(0).start(), false);
+            if(start.duration() >= request.getDuration()){
+                availableTimeRanges.add(start);
+            }
+        }
+    }
```

Comment:
This can be moved into a separate method and reused for line 82-91 by passing correct params.